storage/core/src/main/java/com/hortonworks/registries/storage/impl/jdbc/provider/sql/factory/AbstractQueryExecutor.java

```diff
@@ -200,6 +195,63 @@ public void setStorableFactory(StorableFactory storableFactory) {
     }
 
 
+    // =============== Methods for supporting transaction ===============
+
+
+    @Override
+    public void beginTransaction() throws StorageException {
+        try {
+            if (!transactionBookKeeper.hasActiveTransaction(Thread.currentThread().getId())) {
+                log.debug(" --- Begin transaction for thread id : {} --- ", Thread.currentThread().getId());
+                Connection connection = connectionBuilder.getConnection();
+                log.debug("Opened connection {}", connection);
+                activeConnections.add(connection);
+                connection.setAutoCommit(false);
+                transactionBookKeeper.addTransaction(Thread.currentThread().getId(), connection);
+            } else {
+                log.debug(" --- Reusing transaction for thread if : {} --- ", Thread.currentThread().getId());
+                transactionBookKeeper.incrementNestedTransactionCount(Thread.currentThread().getId());
+            }
+        } catch (SQLException e) {
+            throw new StorageException("Failed to start transaction", e);
+        }
+    }
+
+    @Override
+    public void rollbackTransaction() throws StorageException {
+        Connection connection = transactionBookKeeper.getConnection(Thread.currentThread().getId());
```

Comment:
You need to handle nested transactions here. Rollback should rollback the current connection effectively even when other nested transactions call commit.  
This can be done in the last end-transaction call which can be commit/rollback by maintaining effective state of the transaction in `TransactionBookKeeper`.